sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,376 @@
+    }
+  }
+
+  /**
+   * Unique identifier for a display metadata item within a {@link PTransform} instance.
+   * Identifiers are composed of the key the are registered with and a namespace generated from
+   * the class of the component which registered the item.
+   *
+   * <p>Display metadata registered with the same key from different components will have different
+   * namespaces and will thus both be represented in the composed {@link DisplayData}. If a
+   * single component registers multiple metadata items with the same key, only the most recent
+   * item will be retained; previous versions are discarded.
+   */
+  @AutoValue
+  abstract static class Identifier {
+    static Identifier of(String namespace, String key) {
+      return new AutoValue_DisplayData_Identifier(namespace, key);
+    }
+
+    abstract String getNamespace();
+
+    abstract String getKey();
+  }
+
+  /**
+   * Display metadata type.
+   */
+  static enum Type {
+    STRING {
+      @Override
+      String serializeJsonString(Object value) {
+        return (String) value;
+      }
+    },
+    INTEGER {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString((long) value);
+      }
+    },
+    FLOAT {
+      @Override
+      String serializeJsonString(Object value) {
+        return Double.toString((Double) value);
+      }
+    },
+    TIMESTAMP() {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString(((Instant) value).getMillis());
+      }
+    },
+    DURATION {
+      @Override
+      String serializeJsonString(Object value) {
+        return Long.toString(((Duration) value).getMillis());
+      }
+    },
+    JAVA_CLASS {
+      @Override
+      String serializeJsonString(Object value) {
+        return ((Class<?>) value).getName();
+      }
+    };
+
+    /**
+     * Serialize the display metadata value into a String which can be submitted as JSON
+     * to a UI or diagnostic dashboard.
+     *
+     * <p>Internal-only. Value objects can be safely cast to the expected Java type.
+     */
+    abstract String serializeJsonString(Object value);
+  }
+
+  private static class InternalBuilder implements ItemBuilder {
+    private final Map<Identifier, Item<?>> entries;
+    private final Set<Object> visited;
+
+    private String currentNs;
+    private Item<?> currentItem;
+    private Identifier currentKey;
+
+    private InternalBuilder() {
+      this.entries = Maps.<Identifier, Item<?>>newHashMap();
+      this.visited = Sets.newHashSet();
+    }
+
+    private static InternalBuilder forRoot(HasDisplayData instance) {
+      InternalBuilder builder = new InternalBuilder();
+      builder.include(instance);
+      return builder;
+    }
+
+    @Override
+    public Builder include(HasDisplayData subComponent) {
+      checkNotNull(subComponent);
+      boolean newComponent = visited.add(subComponent);
```

Comment:
I believe this assumes a `hashCode` on instances of `HasDisplayData`. We should document that.